storey/flow.py

```diff
@@ -666,24 +655,52 @@ async def _do(self, event):
             await self._terminate()
             return await self._do_downstream(_termination_obj)
 
-        key = None
-        if self._group_by_key:
-            key = self.key_extractor(event.body) if self.key_extractor else event.key
-
+        key = self._get_event_key(event)
+        
         if len(self._batch[key]) == 0:
             self._batch_time[key] = event.time
 
         if self._timeout_secs and self._timeout_task is None:
-            self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit(key))
+            self._timeout_task = asyncio.get_running_loop().create_task(self._sleep_and_emit())
 
         self._event_count[key] = self._event_count[key] + 1
         self._batch[key].append(self._event_to_batch_entry(event))
 
         if self._event_count[key] == self._max_events:
-            if self._timeout_task is not None and not self._timeout_task.cancelled():
+            if key == self._timeout_task_key and self._timeout_task and not self._timeout_task.cancelled():
                 self._timeout_task.cancel()
+                self._timeout_task_key = None
+            await self._emit_batch(key)
+
+    async def _sleep_and_emit(self):
+        while self._batch:
+            key = next(iter(self._batch.keys()))
+            time_delta = datetime.now(timezone.utc) - self._batch_time[key]
+            delta_seconds = time_delta.total_seconds()
+            if delta_seconds < self._timeout_secs:
+                self._timeout_task_key = key
+                await asyncio.sleep(self._timeout_secs - delta_seconds)
             await self._emit_batch(key)
 
+        self._timeout_task.cancel()
```

Comment:
This **is** the `_timeout_task` so you don't need to cancel it i guess just reset the timeout_key and finish the function.